File: src/com/wang/Game2dEngine/action/VelocityYByAction.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.wang.Game2dEngine.action;

import static java.lang.Math.abs;

/**
 *
 * @author ricolwang
 */
public class VelocityYByAction extends VelocityAction
{

    double velocityYBy;
    float velocityYByDuration;
    double velocityYBySpeed;
    double velocityYByCurrent;

    public VelocityYByAction()
    {
        this.velocityYBy = 0;
        this.velocityYByDuration = 0;
        this.velocityYBySpeed = 0;
        this.velocityYByCurrent = 0;
    }

    public void velocityYBy(double x, float duration)
    {
        if (duration <= 0)
        {
            duration = 0;
            bImmediately = true;
        }
        this.velocityYBy = x;
        this.velocityYByDuration = abs(duration * 1000);
        if (!bImmediately)
        {
            this.velocityYBySpeed = x / velocityYByDuration;
        }
        this.velocityYByCurrent = 0;
    }

    @Override
    public void perform(double runningTime)
    {
        if (bImmediately)
        {
            this.theSprite.setVelocityY(this.theSprite.getVelocityY() + velocityYBy);
            bComplete = true;
        }

        if (bComplete)
        {
            return;
        }

        if (abs(this.velocityYBy) > Action.MINIMUM)
        {
            double x = this.theSprite.getVelocityY();
            double value = this.velocityYBySpeed * runningTime;
            this.velocityYByCurrent += value;

            if (this.velocityYBySpeed > 0)
            {
                if (this.velocityYByCurrent > this.velocityYBy)
                {
                    bComplete = true;
                } else
                {
                    x += value;
                }
            } else
            {
                if (this.velocityYByCurrent < this.velocityYBy)
                {
                    bComplete = true;
                } else
                {
                    x += value;
                }
            }
            this.theSprite.setVelocityY(x);
        } else
        {
            bComplete = true;
        }

    }
}

File: src/com/wang/Game2dEngine/action/VelocityYByAction.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.wang.Game2dEngine.action

/**
 * @author ricolwang
 */
class VelocityYByAction : VelocityAction()
{
    var velocityYBy = 0.0
    var velocityYByDuration = 0f
    var velocityYBySpeed = 0.0
    var velocityYByCurrent = 0.0
    fun velocityYBy(y: Double, duration: Float)
    {
        var duration = duration
        if (duration <= 0)
        {
            duration = 0f
            bImmediately = true
        }
        velocityYBy = y
        velocityYByDuration = Math.abs(duration * 1000)
        if (!bImmediately)
        {
            velocityYBySpeed = y / velocityYByDuration
        }
        velocityYByCurrent = 0.0
    }

    override fun perform(runningTime: Double)
    {
        if (bImmediately)
        {
            theSprite!!.velocityY = theSprite!!.velocityY + velocityYBy
            bComplete = true
        }
        if (bComplete)
        {
            return
        }
        if (Math.abs(velocityYBy) > Action.Companion.MINIMUM)
        {
            var x = theSprite!!.velocityY
            val value = velocityYBySpeed * runningTime
            velocityYByCurrent += value
            if (velocityYBySpeed > 0)
            {
                if (velocityYByCurrent > velocityYBy)
                {
                    bComplete = true
                } else
                {
                    x += value
                }
            } else
            {
                if (velocityYByCurrent < velocityYBy)
                {
                    bComplete = true
                } else
                {
                    x += value
                }
            }
            theSprite!!.velocityY = x
        } else
        {
            bComplete = true
        }
    }
}